### PR TITLE
Fix local node connection issues + better errors

### LINKED ISSUE #53: Customized nodeHost does not work

### PR DESCRIPTION
**Pull request by mvaivre**
- Auto append http protocol to custom node URL
- Return human-readable errors
- Don't throw error on selfClique call issues (previous pseudo health-check)

Fixes #53

### CODE REVIEW

**Comment by polarker** [comment]
@mvaivre did you test it with host or ip different than `127.0.0.1` and `localhost` ?

**Comment by mvaivre** [comment]
> @mvaivre did you test it with host or ip different than `127.0.0.1` and `localhost` ?

Yes I did. But for now I auto-append "http://" only in front of IPs (IPv4), to avoid miss-configurations. 
Should we do the same for `localhost`?

We'll test the full stack tomorrow morning with @tdroxler.

**Comment by mvaivre** [comment]
Note: @tdroxler has CORS issues when communicating with the local node when not using `localhost` or `127.0.0.1`. This would need to be investigated.

File: src/App.tsx [hunk 1]
```diff
@@ -74,27 +74,19 @@ const App = () => {
...
-          type: 'alert'
-        })
-        setClientIsLoading(false)
-      }
+      setClientIsLoading(true)
+      // Get clients
+      const clientResp = await createClient(settings)
+      setClient(clientResp)
+
+      console.log('Clients initialized.')
```
Comment (nop33) [inline src/App.tsx:82]: Why plural, since it's one client?

File: src/pages/SettingsPage.tsx [hunk 2]
```diff
@@ -203,6 +203,11 @@ const ClientSettings = () => {
   const editSettings = (v: Partial<Settings>) => {
     const newSettings = { ...tempSettings, ...v }
 
+    // Check if we need to append the http:// protocol if an IP or is used
```
Comment (nop33) [inline src/pages/SettingsPage.tsx:206]: > if an IP or is used

❓

File: src/utils/api.ts [hunk 3]
```diff
@@ -1,7 +1,24 @@
...
+
+  if (isHTTPError(e)) {
+    returnedError = e.error.detail
+  } else if (stringifiedError) {
+    returnedError = stringifiedError
+  } else {
+    returnedError = defaultErrorMessage
+  }
+
+  return returnedError
```
Comment (nop33) [inline src/utils/api.ts:23]: We could simplify the code to this:

```js
export const getHumanReadableError = (e: unknown, defaultErrorMessage: string) => {
  const stringifiedError = tryGetString(e)

  if (isHTTPError(e)) {
    return e.error.detail
  } else if (stringifiedError) {
    return stringifiedError
  }

  return defaultErrorMessage
}
```

or if you feel fancy, this:

```js
export const getHumanReadableError = (e: unknown, defaultErrorMessage: string) => {
  const stringifiedError = tryGetString(e)
  
  return isHTTPError(e) ? e.error.detail : stringifiedError || defaultErrorMessage
}
```

No strong preference, I leave it up to you :)

File: src/utils/types.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const tryGetString = (obj: any): string | null => {
+  if (typeof obj.toString === 'function') return obj.toString()
+  else return null
+}
```
Comment (nop33) [inline src/utils/types.ts:5]: I personally find the following version simpler, but again, no strong opinion here:

```js
export const tryGetString = (obj: any): string | null => (
  typeof obj.toString === 'function' ? obj.toString() : null
)
```

**Review by nop33 — CHANGES_REQUESTED** [review]

File: src/utils/api.ts (hunk 3 above)
Comment (mvaivre) [reply src/utils/api.ts:23]: Let's go with the ternary version (I hesitated, went for the explicit version, but let's change my mind ;)). I'll make the change.

File: src/utils/types.ts (hunk 4 above)
Comment (nop33) [reply src/utils/types.ts:5]: Yeah, it takes a while to get used to the "no return" statements. Maybe this is simpler then:

```js
export const tryGetString = (obj: any): string | null => {
  return typeof obj.toString === 'function' ? obj.toString() : null
}
```

but anyway, let's not waste more time on this :)

File: src/App.tsx (hunk 1 above)
Comment (mvaivre) [reply src/App.tsx:82]: The main backend "client" is actually an aggregation of the cliqueClient (node) and the ExplorerClient (explorer-backend).

File: src/utils/types.ts (hunk 4 above)
Comment (mvaivre) [reply src/utils/types.ts:5]: It's good to agree on common patterns. Why not go for your last proposition which seems like a good compromise

File: src/pages/SettingsPage.tsx (hunk 2 above)
Comment (mvaivre) [reply src/pages/SettingsPage.tsx]: "or localhost". Fixed.

File: src/App.tsx [hunk 5]
```diff
@@ -74,27 +74,19 @@ const App = () => {
...
-
-        console.log('Clients initialized.')
-
-        setSnackbarMessage({
-          text: `Connected to Alephium's Node "${settings.nodeHost}"!`,
-          type: 'info',
-          duration: 4000
-        })
-        setClientIsLoading(false)
-      } catch (e) {
```
Comment (nop33) [inline src/App.tsx:91]: How do we handle wrong network settings now that you removed the `catch`?

Comment (mvaivre) [reply src/App.tsx:91]: It's the outcome of the whole discussion we had on Slack [here](https://alephium.slack.com/archives/C026MHDJD2N/p1637071135044300) and [here](https://alephium.slack.com/archives/C026MHDJD2N/p1637139773058300) and described in the linked task [here](https://github.com/alephium/alephium-wallet/issues/53#issuecomment-971402782).

Returning selfClique errors could be misleading (it has been, for some users), as we previously considered it as a pseudo-health check. 

- For now, we decided to display an error if an TX cannot be send successfully, indicating a bad configuration or network issue.
- When the user update the settings, we simply print a "Settings saved" message.

(BTW, this shows that brainstorming on Slack is probably not the best solution, as the detailed outcomes are easily lost.)

**Comment by mvaivre** [comment]
@nop33 is your change request still on? 
If not, I believe we could merge this one and go ahead a release a new version.

File: src/App.tsx (hunk 5 above)
Comment (nop33) [reply src/App.tsx:91]: sorry, completely forgot about this. Resolving.

**Review by nop33 — APPROVED** [review]

**Pull request merged**